File: src/paper.py
import os
import json

class Paper:
    def __init__(self, data):
        self.data = data
        self.doi = data.get("doi", "")
        self.data["abstract"] = self._get_abstract()

        year = self.get("publication_year")
        journal = self.get("primary_location").get("source").get("display_name", "")
        self.data["year_journal"] = f"{year} · {journal}" if year and journal else ""

        authors = self.get("authorships")
        self.data["authors_joined"] = ", ".join(author["author"]["display_name"] for author in authors)

        self.valid = bool(self.data.get("abstract")) and bool(self.data.get("authorships"))

    def get(self, key, default=None):
        """Get a value from the paper data, returning default if not found."""
        return self.data.get(key, default)

    def _get_abstract(self):
        abstract_inverted_index = self.data.get("abstract_inverted_index")
        if abstract_inverted_index:
            words_positions = []
            for word, positions in abstract_inverted_index.items():
                for pos in positions:
                    words_positions.append((pos, word))
            words_positions.sort()
            abstract = " ".join(word for _, word in words_positions)
        else:
            abstract = ""
        return abstract

File: web/src/domain/paper.ts
import type { OpenAlexWork, PaperViewModel } from '../types'

const DEFAULT_DOI = '10.1038/s41586-020-2649-2'

const toDoiUrl = (doi?: string) => {
  if (!doi) {
    return ''
  }
  return doi.startsWith('http') ? doi : `https://doi.org/${doi}`
}

const reconstructAbstract = (index?: Record<string, number[]>) => {
  if (!index) {
    return ''
  }
  const tokens: Array<{ pos: number; word: string }> = []
  Object.entries(index).forEach(([word, positions]) => {
    positions.forEach((pos) => tokens.push({ pos, word }))
  })
  tokens.sort((a, b) => a.pos - b.pos)
  return tokens.map((entry) => entry.word).join(' ')
}

const formatYearJournal = (work: OpenAlexWork) => {
  const year = work.publication_year
  const journal = work.primary_location?.source?.display_name ?? ''
  if (!year || !journal) {
    return ''
  }
  return `${year} · ${journal}`
}

const joinAuthors = (work: OpenAlexWork) => {
  const authors = work.authorships ?? []
  return authors
    .map((authorship) => authorship.author?.display_name)
    .filter(Boolean)
    .join(', ')
}

export const toPaperViewModel = (work: OpenAlexWork): PaperViewModel => {
  const doiUrl = toDoiUrl(work.doi ?? DEFAULT_DOI)
  const title = work.display_name ?? work.title ?? 'Untitled'
  const abstract = reconstructAbstract(work.abstract_inverted_index)
  const yearJournal = formatYearJournal(work)
  const authorsJoined = joinAuthors(work)

  return {
    id: work.id,
    doi: doiUrl,
    title,
    abstract,
    year_journal: yearJournal,
    authors_joined: authorsJoined,
    raw: work,
  }
}
